modules/gapi/src/compiler/passes/transformations.cpp

```diff
@@ -0,0 +1,159 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "precomp.hpp"
+
+#include <ade/util/zip_range.hpp>
+#include <ade/graph.hpp>
+
+#include "api/gcomputation_priv.hpp"
+
+#include "compiler/gmodel.hpp"
+#include "compiler/gmodelbuilder.hpp"
+#include "compiler/passes/passes.hpp"
+#include "compiler/passes/pattern_matching.hpp"
+
+#include <sstream>
+
+namespace cv { namespace gimpl { namespace passes {
+namespace
+{
+// Generates ADE graph from expression-based computation
+std::unique_ptr<ade::Graph> makeGraph(const cv::GComputation& c)
+{
+    std::unique_ptr<ade::Graph> pG(new ade::Graph);
+    ade::Graph& g = *pG;
+
+    cv::gimpl::GModel::Graph gm(g);
+    cv::gimpl::GModel::init(gm);
+    cv::gimpl::GModelBuilder builder(g);
+    auto proto_slots = builder.put(c.priv().m_ins, c.priv().m_outs);
+
+    // Store Computation's protocol in metadata
+    cv::gimpl::Protocol p;
+    std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
+    gm.metadata().set(p);
+
+    return pG;
+}
+
+bool tryToSubstitute(ade::Graph& main,
+                     const std::unique_ptr<ade::Graph>& patternG,
+                     const cv::GComputation& substitute)
+{
+    GModel::Graph gm(main);
+
+    // Note: if there are multiple matches, p must be applied several times (outside-scope work)
+    // 1. fina pattern in graph
+    auto match1 = findMatches(*patternG, gm);
+    if (!match1.ok()) {
+        return false;
+    }
+
+    // 2. build substitute graph inside the main graph
+    cv::gimpl::GModelBuilder builder(main);
+    const auto& proto_slots = builder.put(substitute.priv().m_ins, substitute.priv().m_outs);
+    Protocol p;
+    std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
+
+    // 3. match pattern ins/outs to substitute ins/outs
+    auto match2 = matchPatternToSubstitute(*patternG, main,
+        GModel::Graph(*patternG).metadata().get<Protocol>(), p);
+
+    // in theory, we should always be able to find a match here. if not, we're in half-completed
+    // state where some transformations are already applied - what can we do to handle the situation
+    // better?
```

Comment:
This may be done before 2nd step as they are independent

Reply:
In general, I think assert is enough here because it throws an exception (which likely mean that current graph will be destroyed). So in the most common scenario, there's no problem if some transformations are applied already: an exception throw means we need to gracefully close the application or at most start anew from the beginning

Reply:
Oh, found the transaction API. It seems like it's better to introduce it separately (and may be in a generic fashion): this PR is already big enough since we can't merge it for almost a month

Reply:
I agree